Nos/Service/Relay/RelayService.swift

```diff
@@ -454,29 +454,44 @@ extension RelayService {
         client.write(string: requestString)
     }
     
-    /// Opens a websocket and writes a single message to it. On failure this function will just log the error to the 
-    /// console.
+    /// Opens a websocket and writes a single message to it. On failure this
+    /// function will just log the error to the console.
     private func openSocket(to url: URL, andSend message: String) async {
         var urlRequest = URLRequest(url: url)
         urlRequest.timeoutInterval = 10
         let socket = WebSocket(request: urlRequest, compressionHandler: .none)
         
         // Make sure the socket doesn't stay open too long
-        _ = Task(timeout: 10) { socket.disconnect() }
+        let task = Task(timeout: 10) { socket.disconnect() }
         return await withCheckedContinuation({ continuation in
+        
+            var done = false
             socket.onEvent = { (event: WebSocketEvent) in
+                if done {
+                    // We already cleaned up, we don't want to resume the
+                    // continuation twice and any remaining event can be ignored
+                    // now
+                    return
+                }
+                
                 switch event {
                 case WebSocketEvent.connected:
                     socket.write(string: message)
-                    socket.disconnect()
-                case WebSocketEvent.disconnected:
-                    continuation.resume()
                 case WebSocketEvent.error(let error):
                     Log.optional(error, "failed to send message: \(message) to websocket")
+                case WebSocketEvent.disconnected, WebSocketEvent.cancelled:
+                    break
                 default:
                     return
                 }
+                
+
+                done = true
```

Comment:
It looks to me like this Task will just complete right away and not wait for the socket to receive any events? When we set `socket.onEvent = ...` we are just registering a callback to be run whenever the socket receives an event, but then we immediately proceed to say `done = true` and disconnect the socket.

The old code was waiting until we got a connection set up and we had written our message to it before resuming the continuation. I think we need equivalent code to do that. It's tricky because we are bridging old style async code with the new stuff using `withCheckedContinuation`. 

Also I see two issues you are solving for here: one is where the continuation was resumed twice, and one is where the continuation was never resumed. Did you see an actual case where the continuation is never resumed? It looks to me like in the worst case, if continuation.resume is never called after a successful write then our timeout will fire and `socket.disconnect` should cause a `WebSocketEvent.disconnected` event to come through which causes us to call `continuation.resume`.

The reason I ask is because if we only need to solve for the issue where continuation.resume() is called twice we can probably just say `done = true` right before the old calls to `continuation.resume()`. Otherwise we'll need to create a more robust structure around socket.onEvent, like creating a custom class or actor that acts as a delegate object and uses the old NSTimer API to do the timeout or something.

Reply:
So I was seeing two issues. 

1) The most important one was that the continuation was not being called. For some reason the disconnect message sent after writing to the socket was not handled. I also noticed that cancellation messages were ignored, which may be related. Maybe a cancellation sent from the relay after a client disconnects was hitting first? before the client disconnected? In any case, as we only care about is writing and thenwe are done, I though that the safest path was to ensure that for all the paths we always end disconnection and resuming the async func.

Notice that we do wait as before. We only advance if the socket is got disconnected, cancelled, or if we did write to it. In all other cases we wait as before. Notice that the default case has a `return` which is what keeps the wait and keeps listening to events.

2) After doing this I noticed that we also need to ensure the continuation is not called twice so I added the flag as in the library examples.

Reply:
Ok I downloaded the code and I see that this works. But we are relying on the fact that we always get the `connected` event first. Maybe that's a safe assumption to make, but there are lots of other event types like `.ping`, `.pong`, and `.viabilityChanged` that I don't fully understand. In particular `.viabilityChanged` is sometimes called on connection and we actually treat it the same way as `.connected` in `RelayService.didReceive(event: WebSocketEvent, client: WebSocketClient)` so I'm a little worried that some relays might send that event instead of `.connected` first.

I guess if we change `case .connected:` to `case .connected, .viabilityChanged` I will feel a lot better about it.

Reply:
Implemented in https://github.com/planetary-social/nos/pull/1071/commits/ba3dd2bf9b666d984a2b674cac6921b2b4391471